### ISSUE TITLE
SC2039: does not trigger for dash (does for POSIX sh)

### ISSUE THREAD

**Issue by kevinoid** [koalaman/shellcheck#2322]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2039
- My shellcheck version (`shellcheck --version` or "online"): online and 0.7.1
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/dash
set -o pipefail
```

#### Here's what shellcheck currently says:

Nothing

#### Here's what I wanted or expected to see:

    Line 2:
    set -o pipefail
           ^-- SC3040: In POSIX sh, set option pipefail is undefined.

as it does for

```sh
#!/bin/sh
set -o pipefail
```

#### Discussion

Although lack of support for `set -o pipefail` in dash was one of the original motivations in #990, and is noted on the [SC3040 wiki page](https://github.com/koalaman/shellcheck/wiki/SC3040), it appears that `set -o pipefail` is not currently triggered for dash.

**Note:** [`set -o pipefail` has been accepted for POSIX 202x](https://austingroupbugs.net/view.php?id=789) and it is becoming more widely supported (e.g. bash, ksh, zsh, [busybox ash](https://github.com/koalaman/shellcheck/issues/990#issuecomment-329237300), [FreeBSD sh](https://bugs.freebsd.org/224270) - although [still not dash as of July 2020](https://lore.kernel.org/dash/20200701015226.GA6839@gondor.apana.org.au/)).  At some point it may be worth limiting SC3040 to only shells known to lack support.

Thanks for considering,
Kevin

**Comment by kurahaupo** [comment]
There will continue to be old versions of shells that can't or won't be upgraded until the devices that run them become landfill. So, for at least the next 10 years, this will be a matter of recognizing multiple versions of each shell.

I suggest that ShellCheck could add a <tt>--posix=<i>nnnn</i></tt> option and use whichever version of POSIX was ratified in or before the year _nnnn_, with a default of the current year minus 5. (If _nnnn_ contains "x" then take it as "9" so that this comparison makes sense.)
